### PR TITLE
New approach to unification

### PR DESCRIPTION
**Pull request by paf31**
This is step one in the move toward a new typechecker approach.

- Break up `Control.Monad.Unify`
- Use `mtl` style type signatures
- Walk the substitution _before_ unification instead of the weird old approach where substitution updates could trigger other unifications.

### CODE REVIEW

**Comment by paf31** [comment]
I think this might fix #1429 but I need to test more.

File: src/Language/PureScript/TypeChecker/Kinds.hs [hunk 1]
```diff
@@ -48,177 +46,213 @@ import Language.PureScript.Names
...
+  sub <- gets checkSubstitution
+  go (substituteKind sub k1) (substituteKind sub k2)
+  where
+  go (KUnknown u1) (KUnknown u2) | u1 == u2 = return ()
+  go (KUnknown u) k = solveKind u k
+  go k (KUnknown u) = solveKind u k
+  go Star Star = return ()
+  go Bang Bang = return ()
+  go (Row k1') (Row k2') = go k1' k2'
+  go (FunKind k1' k2') (FunKind k3 k4) = do
```
Comment (jdegoes) [inline src/Language/PureScript/TypeChecker/Kinds.hs]: This is a perfect example of where shadowing is a good thing, and can prevent you from accidentally referencing stale parameters; i.e. you can bind `k1` and `k2` to make sure you cannot reference them from inside the definition of the function. Albeit in this case, I manually verified the code so it looks good to me. :+1:

Comment (paf31) [inline src/Language/PureScript/TypeChecker/Kinds.hs]: Yeah this is less than ideal, I'll tidy it up, thanks!

**Comment by jdegoes** [comment]
While reviewing the code, I also couldn't help notice the (usual) duplication between kind and type inference (`KUnknown Int` / `TUnknown Int`, `unifyKinds` / `unifyTypes`, etc.), and found [a proposal to unify kind and type inference in UHC](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=0CB0QFjAAahUKEwiv_7fklvLIAhULOj4KHeCNAag&url=http%3A%2F%2Fwww.cs.uu.nl%2Fdocs%2Fvakken%2Fts%2FPresentations%2FRemkoJeroen.ppt&usg=AFQjCNHsFL-ebz5b_prI_GWIUktbtcBjQQ&bvm=bv.106379543,d.cWw). Too big a change for this PR but I thought I'd at least link it for future reference.

**Comment by paf31** [comment]
Yes, this PR actually is a move backwards in that regard, sort of. The previous code used to share a constraint solving module, but this pulls that apart. The reasons are that a) it makes the code much simpler, b) the old approach required orphan instances and c) I plan for the two implementations to differ more in the future, as I add new types of constraints and change the approach in the type checker portion to rely less on unification.

**Comment by jdegoes** [comment]
Yeah, if they're going to be diverging, makes sense. :+1:

**Comment by paf31** [comment]
This should build now.

**Comment by paf31** [comment]
This fixes #1429 technically, in that it doesn't cause the compiler to spin any more, but it also allows an invalid program through (duplicate rows). I'm personally fine with that, since the duplicate rows check needs attention anyway, and might even be removed soon.

**Comment by garyb** [comment]
I haven't reviewed this yet, but just thought I'd report I've had no problems with it either, after a day of heavy usage.

**Comment by garyb** [comment]
:shipit:

**Comment by paf31** [comment]
Thanks!

**Pull request merged**